### PR TITLE
init

### PR DESCRIPTION
**Pull request by menshibin**

### CODE REVIEW

File: nodejs/src/client/wsClient.ts [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    WSFetchBlockResponse,
+    WSQueryResponse,
+    WSFetchResponse,
+} from './wsResponse';
+import { ReqId } from '../common/reqid';
+
+
+export class WsClient {
+    private _wsConnector?: WebSocketConnector;
+    private _req_id = 1000000;
```
Comment (huskar-t) [inline nodejs/src/client/wsClient.ts:17]: unsued

File: nodejs/src/client/wsClient.ts [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    private _timeout?:number | undefined | null;
+    private _url:URL;
```
Comment (huskar-t) [inline nodejs/src/client/wsClient.ts:19]: readonly?

File: nodejs/package.json [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    "json-bigint": "^1.0.0",
+    "websocket": "^1.0.34"
+  },
+  "devDependencies": {
+    "@parcel/packager-ts": "^2.7.0",
+    "@parcel/transformer-typescript-types": "^2.7.0",
+    "@types/jest": "^29.2.1",
+    "@types/json-bigint": "^1.0.1",
+    "@types/node": "^18.0.0",
+    "@types/websocket": "^1.0npm install async-mutex.5",
```
Comment (huskar-t) [inline nodejs/package.json:50]: wrong version

File: nodejs/src/common/taosResult.ts [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+
+export function readVarchar(dataBuffer: ArrayBuffer, colDataHead: number, length: number): string {
+    let data = "";
+    let buff = dataBuffer.slice(colDataHead, colDataHead + length)
+    data += new TextDecoder().decode(buff)
+    return data;
+}
+
+export function readNchar(dataBuffer: ArrayBuffer, colDataHead: number, length: number): string {
+    let decoder = new TextDecoder();
```
Comment (huskar-t) [inline nodejs/src/common/taosResult.ts:424]: unused

File: nodejs/src/common/wsError.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            (this as any).__proto__ = new.target.prototype;
+        }
+    }
+}
+export class WebSocketQueryError extends TDWebSocketClientError { }
+export class WebSocketInterfaceError extends TDWebSocketClientError {}
+export class WebSocketQueryInterFaceError extends WebSocketInterfaceError{}
+export class TaosResultError extends TDWebSocketClientError{};
+export class TaosError extends TDWebSocketClientError{};
+
```
Comment (huskar-t) [inline nodejs/src/common/wsError.ts:22]: Unnecessary semicolon

File: nodejs/src/common/wsError.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+export enum ErrorCode {
+    ERR_INVALID_PARAMS = 100,
+    ERR_INVALID_URL = 101,
+    ERR_WS_NO_CALLBACK = 102,
+    ERR_INVALID_MESSAGE_TYPE = 103,
+    ERR_WEBSOCKET_CONNECTION = 104,
+    ERR_WEBSOCKET_QUERY_TIMEOUT = 105,
+    ERR_INVALID_AUTHENTICATION = 106,
+    ERR_UNSPPORTED_TDENGINE_TYPE = 107,
```
Comment (huskar-t) [inline nodejs/src/common/wsError.ts:31]: typo? unsupported?

File: nodejs/src/sql/wsRows.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            return false;
+        }
+        return true;
+    }
+
+    private async getBlockData():Promise<TaosResult> {
+        try {
+            let wsFetchResponse = await this._wsClient.fetch(this._wsQueryResponse);
+            console.log("[wsQuery.execute.wsFetchResponse]==>\n", wsFetchResponse)
+            if (wsFetchResponse.completed == true) {
```
Comment (huskar-t) [inline nodejs/src/sql/wsRows.ts:42]: Can be simplified to wsFetchResponse.completed

File: nodejs/src/sql/wsRows.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        
+        let data = this._taosResult.GetData();
+        if (this._taosResult && data != null) {
+            if (data && Array.isArray(this._taosResult.GetData()) && data.length > 0) {
+                return true;
+            }
+        }
+
+        this._taosResult = await this.getBlockData();
+        if (this._taosResult.GetData() == null) {
```
Comment (huskar-t) [inline nodejs/src/sql/wsRows.ts:32]: 'if' statement can be simplified

File: nodejs/src/sql/wsSql.ts [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        try {
+            let wsQueryResponse:WSQueryResponse = await this._wsClient.exec(this.getSql(sql, reqId, action));
+            let taosResult = new TaosResult(wsQueryResponse);
+            if (wsQueryResponse.is_update == true) {
+                return taosResult;
+            } else {
+                try{
+                    while (true) {
+                        let wsFetchResponse = await this._wsClient.fetch(wsQueryResponse)
+                        if (wsFetchResponse.completed == true) {
```
Comment (huskar-t) [inline nodejs/src/sql/wsSql.ts:128]: Can be simplified to wsFetchResponse.completed

File: nodejs/src/tmq/config.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
+import { TMQConstants } from "./constant";
+
+export class TmqConfig {
+    // req_id: number;
+    url: URL;
+    user: string;
+    password: string;
+    group_id: string;
+    clien_id: string;
```
Comment (huskar-t) [inline nodejs/src/tmq/config.ts:9]: typo?

File: nodejs/src/tmq/tmqResponse.ts [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        }
+        for (let i in this.positions) {
+            topicPartitions[i].offset = this.positions[i]
+        }
+        return topicPartitions;
+    } 
+    
+}
+
+export class CommitedResp extends PartitionsResp {
```
Comment (huskar-t) [inline nodejs/src/tmq/tmqResponse.ts:153]: typo?

File: nodejs/src/tmq/wsTmq.ts [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+                req_id    : ReqId.getReqID(reqId),
+                blocking_time  :timeoutMs
+            },
+        };
+        return new Promise(async (resolve, reject) => {
+            try {
+                var taosResults: Map<string, TaosResult> = new Map();
+                let resp = await this._wsClient.exec(JSON.stringify(queryMsg), false);
+                let pollResp = new WsPollResponse(resp)
+                if (pollResp.have_message == false || pollResp.message_type != TMQMessageType.ResDataType) {
```
Comment (huskar-t) [inline nodejs/src/tmq/wsTmq.ts:382]: Can be simplified to !pollResp.have_message

File: nodejs/src/tmq/wsTmq.ts [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+                    resolve(taosResults);
+                } else {        
+                    // let count = 0
+                    // let startTime = new Date().getTime();
+                    while (true) {
+                        // count++
+                        let fetchResp = await this.fetch(pollResp)
+                        if (fetchResp.completed || fetchResp.rows == 0) {
+                            let currTime = new Date().getTime();
```
Comment (huskar-t) [inline nodejs/src/tmq/wsTmq.ts:391]: unused

File: nodejs/src/client/wsConnectorPool.ts [hunk 14]
```diff
@@ -0,0 +1,112 @@
+import { Mutex } from "async-mutex";
+import { WebSocketConnector } from "./wsConnector";
+import { ErrorCode, TDWebSocketClientError } from "../common/wsError";
+
+const mutex = new Mutex();
+export class WebSocketConnectionPool {
+    private static _instance?:WebSocketConnectionPool;
+    private pool: Map<string, WebSocketConnector[]> = new Map();
+    private  _connectionCount: number;
+    private  _maxConnections: number;
```
Comment (huskar-t) [inline nodejs/src/client/wsConnectorPool.ts:10]: readonly?

File: nodejs/src/sql/wsRows.ts [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+        } 
+        return undefined;
+    }
+
+    Close() {
+        if (this._isClose) {
+            return
+        }
+        this._isClose = true
+        this._wsClient.freeResult(this._wsQueryResponse)
```
Comment (huskar-t) [inline nodejs/src/sql/wsRows.ts:79]: need .then()?

File: nodejs/src/sql/wsRows.ts [hunk 16]
```diff
@@ -0,0 +1,82 @@
+import { TDengineMeta, TaosResult } from '../common/taosResult';
+import { TaosResultError } from '../common/wsError';
+import { WSQueryResponse } from '../client/wsResponse';
+import { WsClient } from '../client/wsClient';
+
+export class WSRows {
+    private _wsClient: WsClient;
+    private _wsQueryResponse: WSQueryResponse;
```
Comment (huskar-t) [inline nodejs/src/sql/wsRows.ts:8]: readonly?

File: nodejs/example/basicSql.ts [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+        wsRows = await wsSql.Query('select * from meters', reqId++);
+        let meta = wsRows.GetMeta()
+        console.log("wsRow:meta:=>", meta);
+
+        while (await wsRows.Next()) {
+            let result = await wsRows.GetData();
+            console.log('queryRes.Scan().then=>', result);
+        }
+        await wsRows.Close()
```
Comment (huskar-t) [inline nodejs/example/basicSql.ts:45]: wrRows.Close is a synchronous function

**Review by huskar-t — CHANGES_REQUESTED** [review]
see my comments

File: nodejs/src/common/taosResult.ts [hunk 18]
```diff
@@ -0,0 +1,466 @@
...
+        return this._meta;
+    }
+
+    public GetPrecision():number | null | undefined {
+        return this._precision;
+    }
+    public GetTotalTime() {
+        return this._totalTime;
+    }
+    public  AddtotalTime(totalTime:number) {
```
Comment (huskar-t) [inline nodejs/src/common/taosResult.ts:103]: AddtotalTime -> AddTotalTime

File: nodejs/src/sql/wsSql.ts [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+                    await this.execute(`use ${database}`)
+                }
+                resolve(this)
+            } catch(e) {
+                reject(e)
+            }
+        })
+    }
+
+    async execute(sql: string, reqId?: number, action:string = 'query'): Promise<TaosResult> {
```
Comment (sheyanjie-qq) [inline nodejs/src/sql/wsSql.ts:118]: schemaless 不支持sql

**Comment by huskar-t** [comment]
need to format and add automatic tests

File: nodejs/src/stmt/wsParams.ts [hunk 20]
```diff
@@ -0,0 +1,465 @@
...
+                bitmapBuffer.setInt8(charOffset, nullVal);                    
+            }
+        }
+
+        this._dataTotalLen += arrayBuffer.byteLength; 
+        this._params.push(new ColumnInfo([TDengineTypeLength['TIMESTAMP'] * params.length, arrayBuffer], TDengineTypeCode['TIMESTAMP'], TDengineTypeLength['TIMESTAMP']));  
+    }
+
+
+    private encodeColumnNumbers(params:any[], dataType:string = 'number', typeLen:number, columnType:number):[number, ArrayBuffer] {
```
Comment (sheyanjie-qq) [inline nodejs/src/stmt/wsParams.ts:231]: 方法名好像只处理 number 类型

File: nodejs/src/tmq/wsTmq.ts [hunk 21]
```diff
@@ -0,0 +1,470 @@
...
+                var taosResults: Map<string, TaosResult> = new Map();
+                let resp = await this._wsClient.exec(JSON.stringify(queryMsg), false);
+                let pollResp = new WsPollResponse(resp)
+                if (pollResp.have_message == false || pollResp.message_type != TMQMessageType.ResDataType) {
+                    resolve(taosResults);
+                } else {        
+                    // let count = 0
+                    // let startTime = new Date().getTime();
+                    while (true) {
+                        // count++
```
Comment (sheyanjie-qq) [inline nodejs/src/tmq/wsTmq.ts:388]: 无用代码都应该删除，不要注释掉

File: nodejs/test/bulkPulling/queryTables.test.ts [hunk 22]
```diff
@@ -0,0 +1,298 @@
...
+                expect(meta.name).toBe(expectMeta[index].name)
+                expect(meta.type).toBe(expectMeta[index].type)
+                expect(meta.length).toBe(expectMeta[index].length)
+                ////   console.log(meta);
+            })
+
+            for (let i = 0; i < actualData.length; i++) {
+                actualData[i].forEach((d, index) => {
+                    if (expectMeta[index].name == 'geo' || expectMeta[index].name == 'vbinary') {
+                        expect(d).toBeTruthy()
```
Comment (sheyanjie-qq) [inline nodejs/test/bulkPulling/queryTables.test.ts:128]: 为何这两种类型没有仔细检查？

File: nodejs/test/bulkPulling/schemaless.test.ts [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+        conf.SetDb('power')
+        let wsSchemaless = await WsSql.Open(conf)
+        expect(wsSchemaless.State()).toBeGreaterThan(0)
+        await wsSchemaless.SchemalessInsert([influxdbData], SchemalessProto.InfluxDBLineProtocol, Precision.NANO_SECONDS, 0);
+        await wsSchemaless.SchemalessInsert([telnetData], SchemalessProto.OpenTSDBTelnetLineProtocol, Precision.SECONDS, 0);
+        await wsSchemaless.SchemalessInsert([jsonData], SchemalessProto.OpenTSDBJsonFormatProtocol, Precision.SECONDS, 0);
+        wsSchemaless.Close();
+    });
+
+    test('normal wsSql insert', async() => {
```
Comment (sheyanjie-qq) [inline nodejs/test/bulkPulling/schemaless.test.ts:72]: 这里都没有check

**Pull request merged**